### PR TITLE
API-18233-Add-claimPhaseDates

### PR DESCRIPTION
**Pull request by stiehlrod**
## Description of change
Adds claimPhaseDates to GET /veterans/{veteranId}/claims with the following attributes:

- phaseChangeDate

Adds claimPhaseDates to GET /veterans/{veteranId}/claims/{id} with the following attributes:

- currentPhaseBack
- everPhaseBack
- latestPhaseType
- phaseChangeDate

## Original issue(s)
[API-18233](https://vajira.max.gov/browse/API-18233)

### CODE REVIEW

File: spec/support/schemas/claims_api/v2/veterans/claims/claim.json [hunk 1]
```diff
@@ -37,6 +37,11 @@
       "description": "Date claim was closed",
       "example": "2019-09-04"
     },
+    "currentPhaseBack": {
+      "type": "boolean",
+      "nullable": true,
```
Comment (jw81) [inline spec/support/schemas/claims_api/v2/veterans/claims/claim.json:42]: I'm not sure we want this to be nullable. 🤔 
I believe it should be either `true` or `false`.

File: modules/claims_api/app/controllers/claims_api/v2/veterans/claims_controller.rb [hunk 2]
```diff
@@ -148,17 +149,21 @@ def build_claim_structure(data:, lighthouse_id:, upstream_id:) # rubocop:disable
...
             decision_letter_sent: map_yes_no_to_boolean(
               'decision_notification_sent',
               data[:decision_notification_sent]
             ),
             development_letter_sent: map_yes_no_to_boolean('development_letter_sent', data[:development_letter_sent]),
             documents_needed: map_yes_no_to_boolean('attention_needed', data[:attention_needed]),
             end_product_code: data[:end_prdct_type_cd],
+            ever_phase_back: ever_phase_back(data),
             jurisdiction: data[:regional_office_jrsdctn],
+            latest_phase_type: latest_phase_type(data),
```
Comment (jw81) [inline modules/claims_api/app/controllers/claims_api/v2/veterans/claims_controller.rb:162]: For attributes that will only exist on the response in the `show` endpoint, @acovrig and I have been following a pattern of just appending those attributes [here](https://github.com/department-of-veterans-affairs/vets-api/blob/master/modules/claims_api/app/controllers/claims_api/v2/veterans/claims_controller.rb#L68-#L70).

The way it works (at least in my head) is that this `build_claim_structure()` method should really only focus on building the attributes that are common between the `index` and `show` endpoint responses.  Anything else that needs to be added to the `show` response, should just be appended to the common structure over in the `generate_show_output()` method.

So like in your case, `phase_change_date` is going to exist on both endpoint responses. So yeah, let's add that attribute here in `build_claim_structure()`.  But `current_phase_back`, `ever_phase_back`, and `latest_phase_type` are only going to exist on the `show` response.  So let's add those over in `generate_show_output()`.

That make sense?  You ok with that?

File: modules/claims_api/app/controllers/claims_api/v2/veterans/claims_controller.rb [hunk 3]
```diff
@@ -168,6 +173,46 @@ def build_claim_structure(data:, lighthouse_id:, upstream_id:) # rubocop:disable
...
+        def ever_phase_back(data)
+          return if data[:phase_type_change_ind].nil?
+
+          pt_ind_array = get_phase_type_indicator_array(data)
+          pt_ind_array.first.to_i > pt_ind_array.last.to_i
+        end
+
+        def get_evss_claim(data)
+          evss_id = data[:benefit_claim_id]
+          ClaimsApi::EVSSClaim.new(evss_id: evss_id, data: data)
```
Comment (jw81) [inline modules/claims_api/app/controllers/claims_api/v2/veterans/claims_controller.rb:191]: Hmmm, I'm not sure we're gonna want to do this because one of the big goals with V2 is to completely disconnect from EVSS. (it gets permanently sunset 1-yr from today).  

If we start adding new logic into old objects explicitly named EVSS, that's going to lead to confusion for devs in the future when EVSS doesn't even exist.

There's already a simple service that exists for trying to map the value of a BGS claim `status` attribute to the string values we want our API to provide. ([REF](https://github.com/department-of-veterans-affairs/vets-api/blob/master/modules/claims_api/lib/claims_api/bgs_claim_status_mapper.rb))

Maybe we can encapsulate this new mapping logic you need over in there?

File: modules/claims_api/app/controllers/claims_api/v2/veterans/claims_controller.rb [hunk 4]
```diff
@@ -157,8 +161,9 @@ def build_claim_structure(data:, lighthouse_id:, upstream_id:) # rubocop:disable
             end_product_code: data[:end_prdct_type_cd],
             jurisdiction: data[:regional_office_jrsdctn],
             lighthouse_id: lighthouse_id,
-            max_est_claim_date: data[:max_est_claim_complete_dt],
-            min_est_claim_date: data[:min_est_claim_complete_dt],
+            maxEstClaimDate: data[:max_est_claim_complete_dt],
+            minEstClaimDate: data[:min_est_claim_complete_dt],
```
Comment (jw81) [inline modules/claims_api/app/controllers/claims_api/v2/veterans/claims_controller.rb:165]: Why do we need to switch this to camelCase?

File: modules/claims_api/app/controllers/claims_api/v2/veterans/claims_controller.rb [hunk 5]
```diff
@@ -168,6 +173,49 @@ def build_claim_structure(data:, lighthouse_id:, upstream_id:) # rubocop:disable
...
+        def ever_phase_back(data)
+          return false if data[:benefit_claim_details_dto][:phase_type_change_ind].nil?
+
+          pt_ind_array = get_phase_type_indicator_array(data)
+          pt_ind_array.first.to_i > pt_ind_array.last.to_i
+        end
+
+        def get_evss_claim(data, phase_number)
+          ClaimsApi::BGSClaimStatusMapper.new(data[:benefit_claim_details_dto], phase_number).name_from_phase
+        end
```
Comment (jw81) [inline modules/claims_api/app/controllers/claims_api/v2/veterans/claims_controller.rb:191]: This method is pretty confusing.
It's called "get_evss_claim", but V2 isn't getting claims from EVSS.
The method also isn't actually getting a claim.
It's getting a name value.

File: modules/claims_api/app/models/claims_api/evss_claim.rb [hunk 6]
```diff
@@ -35,6 +35,17 @@ class EVSSClaim
       8 => 'Complete'
```
Comment (jw81) [inline modules/claims_api/app/models/claims_api/evss_claim.rb:35]: Are the changes in this file still needed?

File: modules/claims_api/app/controllers/claims_api/v2/veterans/claims_controller.rb [hunk 7]
```diff
@@ -68,6 +70,7 @@ def generate_show_output(bgs_claim:, lighthouse_claim:) # rubocop:disable Metric
           structure.merge!(errors: get_errors(lighthouse_claim))
           structure.merge!(supporting_documents: build_supporting_docs(bgs_claim))
           structure.merge!(tracked_items: map_bgs_tracked_items(bgs_claim))
+          structure.merge!(build_additional_bgs_attributes(bgs_claim))
```
Comment (jw81) [inline modules/claims_api/app/controllers/claims_api/v2/veterans/claims_controller.rb:73]: This should really be named something more specific like, `build_claim_phase_attributes()`.

**Comment by jw81** [comment]
I was just looking at the AC on the ticket and noticed this (also confirmed with Kayla), but these new attributes need to be added into an attribute named `claimPhaseDates`.

It looks like these new attributes are being added directly into the root of the response.  But it looks like Kayla wanted them added into a new attribute named `claimPhaseDates`.

File: modules/claims_api/app/models/claims_api/auto_established_claim.rb [hunk 8]
```diff
@@ -36,7 +36,8 @@ class AutoEstablishedClaim < ApplicationRecord
 
     EVSS_CLAIM_ATTRIBUTES = %i[date_filed min_est_date max_est_date open waiver_submitted
                                documents_needed development_letter_sent decision_letter_sent
-                               requested_decision va_representative].freeze
+                               requested_decision va_representative current_phase_back ever_phase_back
+                               latest_phase_type phase_change_date].freeze
```
Comment (jw81) [inline modules/claims_api/app/models/claims_api/auto_established_claim.rb:40]: Can you help me understand why this change is needed?

Comment (stiehlrod) [reply modules/claims_api/app/models/claims_api/auto_established_claim.rb:40]: Yes, so this was one of the many changes from Kayla. She wants to provide the actual status or phase from BGS, rather than the condensed phases that we normally use. So in the show, more detail.
The initialize is called from two places now, one just needs the normal phase, the other needs to derive the phase from the `phase_type_change_ind`, which is an int, and the mapper converts that int to a BGS phase.
Does my explanation make sense? Do you have a preferred way to handle this scenario?

Comment (jw81) [reply modules/claims_api/app/models/claims_api/auto_established_claim.rb:40]: Ok.  I guess more specifically, I'm wondering if this change actually changes anything?
From what I can tell, it looks like the only place `EVSS_CLAIM_ATTRIBUTES` is ever referenced in `vets-api` is directly below this ([right here](https://github.com/department-of-veterans-affairs/vets-api/blob/93e17dae18113b0d3df2447224e25a4d5c7814c4/modules/claims_api/app/models/claims_api/auto_established_claim.rb#L44)).  And it looks like all it does is create a method for each attribute that returns `nil`.  (I honestly have no idea why that even exists 😂) If you remove this change, does it affect any behavior or anything in your PR?

Comment (stiehlrod) [reply modules/claims_api/app/models/claims_api/auto_established_claim.rb:40]: omg, when I clicked on `view changes`, it took me somewhere else. No, this is not needed, and I will remove it now!

**Review by jw81 — APPROVED** [review]
